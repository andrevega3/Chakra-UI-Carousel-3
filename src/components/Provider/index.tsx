import React, { createContext, useCallback, useMemo, useState } from "react";

export interface ContextType {
  trackIsActive: boolean;
  setTrackIsActive: (value: boolean) => void;
  multiplier: number;
  sliderWidth: number;
  activeItem: number;
  setActiveItem: (value: number) => void;
  constraint: number;
  initSliderWidth: (width: number) => void;
  itemWidth: number;
  setMultiplier: (value: number) => void;
  setConstraint: (value: number) => void;
  setItemWidth: (value: number) => void;
  positions: number[];
  setPositions: (value: number[]) => void;
  infinite: boolean;
  autoplay: boolean;
}

export const Context = createContext<ContextType | undefined>(undefined);

interface ProviderProps {
  children: React.ReactNode;
  infinite?: boolean;
  autoplay?: boolean;
}

export const Provider: React.FC<ProviderProps> = ({ 
  children,
  infinite = false,
  autoplay = false,
 }) => {
  const [trackIsActive, setTrackIsActive] = useState(false);
  const [multiplier, setMultiplier] = useState(0.35);
  const [sliderWidth, setSliderWidth] = useState(0);
  const [activeItem, setActiveItem] = useState(0);
  const [constraint, setConstraint] = useState(0);
  const [itemWidth, setItemWidth] = useState(0);
  const [positions, setPositions] = useState<number[]>([]);

  const initSliderWidth = useCallback(
    (width: number) => setSliderWidth(width),
    []
  );

  const value = useMemo(
    () => ({
      trackIsActive,
      setTrackIsActive,
      multiplier,
      sliderWidth,
      activeItem,
      setActiveItem,
      constraint,
      initSliderWidth,
      itemWidth,
      setMultiplier,
      setItemWidth,
      setConstraint,
      positions,
      setPositions,
      infinite,
      autoplay,
    }),
    [
      trackIsActive,
      setTrackIsActive,
      multiplier,
      sliderWidth,
      activeItem,
      setActiveItem,
      constraint,
      initSliderWidth,
      itemWidth,
      setMultiplier,
      setItemWidth,
      setConstraint,
      positions,
      setPositions,
      infinite,
      autoplay,
    ]
  );

  return <Context.Provider value={value}>{children}</Context.Provider>;
};
