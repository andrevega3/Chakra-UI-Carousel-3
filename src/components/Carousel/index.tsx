import { useBreakpointValue } from "@chakra-ui/react";
import React, { useContext, useEffect, useMemo } from "react";
import Item from "../Item";
import { Context, ContextType } from "../Provider";
import Slider from "../Slider";
import Track from "../Track";

export interface CarouselPropTypes {
  children: React.ReactNode[];
  gap: number;
}

export const Carousel: React.FC<CarouselPropTypes> = ({
  children,
  gap,
}) => {
  const context = useContext(Context);
  if (!context) {
    console.error("‚ùå Carousel must be wrapped in <Provider>");
    return null;
  }

  const {
    setItemWidth,
    sliderWidth,
    setMultiplier,
    setConstraint,
    itemWidth,
    setPositions,
    multiplier,
    constraint,
    positions,
    infinite,
    autoplay,
  } = context;

  const layout = useBreakpointValue({
    base: { multiplier: 0.65, constraint: 1, divisor: 1 },
    md: { multiplier: 0.5, constraint: 2, divisor: 2 },
    xl: { multiplier: 0.35, constraint: 3, divisor: 3 },
  });

  const mapChildren = useMemo(() => {
    if (!infinite || constraint === 0) return children;

    const result = [
      ...children.slice(-constraint),       // prepended clones
      ...children,
      ...children.slice(0, constraint + 1)  // extended clones after end
    ];

    console.log("üßÆ Rendering indexes:", result.map((_, i) => i));
    return result;
  }, [children, infinite, constraint]);

  useEffect(() => {
    if (!layout) return;

    const newItemWidth = sliderWidth / layout.divisor - gap;

    setItemWidth(newItemWidth);
    setMultiplier(layout.multiplier);
    setConstraint(layout.constraint);
  }, [layout, sliderWidth, gap]);

  useEffect(() => {
    const newPositions = mapChildren.map(
      (_, index) => -Math.abs((itemWidth + gap) * index)
    );

    if (JSON.stringify(newPositions) !== JSON.stringify(positions)) {
      setPositions(newPositions);
      console.log("üìç Updated Positions:", newPositions);
    }
  }, [mapChildren, gap, itemWidth, positions, setPositions]);

  return (
    <Slider gap={gap}>
      <Track>
        {mapChildren.map((child, index) => (
          <Item gap={gap} key={index} index={index}>
            {child}
          </Item>
        ))}
      </Track>
    </Slider>
  );
};