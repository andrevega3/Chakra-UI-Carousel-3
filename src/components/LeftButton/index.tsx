import {
  Box,
  Button,
  ButtonProps
} from "@chakra-ui/react";
import { ChevronLeftIcon } from "lucide-react";
import React, { useContext, ReactElement } from "react";
import { Context, ContextType } from "../Provider";

interface CustomButtonProps extends ButtonProps {
  customIcon?: ReactElement;
}

export const LeftButton = ({
  customIcon,
  ...props
}: CustomButtonProps) => {
  const context = useContext(Context);

  const { setTrackIsActive, activeItem, positions, setActiveItem, constraint, infinite } =
    context as ContextType;

  const handleFocus = () => setTrackIsActive(true);

  const handleDecrementClick = () => {
    setTrackIsActive(true);

    console.log("‚¨ÖÔ∏è LEFT CLICK");
    console.log("activeItem:", activeItem);
    console.log("currentPosition:", positions[activeItem]);

    if (infinite) {
      const total = positions.length;
      const firstReal = constraint;
      const lastReal = total - constraint;

      if (activeItem - 1 < firstReal) {
        console.log("üîÅ Wrapping to END:", lastReal - 1);
        setActiveItem(lastReal - 1); // snap to end clone
      } else {
        setActiveItem(activeItem - 1);
      }
    } else {
      if (activeItem > 0) {
        setActiveItem(activeItem - 1);
      }
    }
  };

  return (
    <Button
      {...props}
      onClick={handleDecrementClick}
      onFocus={handleFocus}
      zIndex={2}
      minW={0}
    >
      {customIcon ? <Box>{customIcon}</Box> : <ChevronLeftIcon size={9} />}
    </Button>
  );
};

