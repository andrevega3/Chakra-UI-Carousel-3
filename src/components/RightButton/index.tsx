import { ChevronRightIcon } from "lucide-react";
import {
  Box,
  Button,
  ButtonProps
} from "@chakra-ui/react";
import React, { useContext, ReactElement } from "react";
import { Context, ContextType } from "../Provider";

interface CustomButtonProps extends ButtonProps {
  customIcon?: ReactElement;
}

export const RightButton = ({
  customIcon,
  ...props
}: CustomButtonProps) => {
  const context = useContext(Context);

  const { setTrackIsActive, activeItem, constraint, positions, setActiveItem, infinite } =
    context as ContextType;

  const handleFocus = () => setTrackIsActive(true);

  const handleIncrementClick = () => {
    setTrackIsActive(true);

    console.log("‚û°Ô∏è RIGHT CLICK");
    console.log("activeItem:", activeItem);
    console.log("currentPosition:", positions[activeItem]);

    if (infinite) {
      const total = positions.length;
      const firstReal = constraint;
      const lastReal = total - constraint;

      if (activeItem + 1 >= lastReal) {
        console.log("üîÅ Wrapping to START:", firstReal);
        setActiveItem(firstReal); // Snap to beginning clone
      } else {
        setActiveItem(activeItem + 1);
      }
    } else {
      if (activeItem < positions.length - constraint) {
        setActiveItem(activeItem + 1);
      }
    }
  };

  return (
    <Button
      {...props}
      onClick={handleIncrementClick}
      onFocus={handleFocus}
      zIndex={2}
      minW={0}
    >
      {customIcon ? <Box>{customIcon}</Box> : <ChevronRightIcon size={9} />}
    </Button>
  );
};

