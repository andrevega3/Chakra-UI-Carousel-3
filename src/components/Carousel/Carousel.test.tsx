import React from "react";
import { Carousel } from ".";
import { render } from "@testing-library/react";
import { DemoComp } from "./Carousel.stories";
import { Provider } from "../Provider";
import { ChakraProvider, defaultSystem } from "@chakra-ui/react"
import { LeftButton } from "../LeftButton";
import { RightButton } from "../RightButton";



describe("renders correctly", () => {
  it("Carousel renders correctly", () => {
    const { asFragment } = render(
        <ChakraProvider value={defaultSystem}>
          <Provider>
            <Carousel gap={3}>
              <DemoComp index={1} />
              <DemoComp index={2} />
              <DemoComp index={3} />
              <DemoComp index={4} />
            </Carousel>
          </Provider>
        </ChakraProvider>
      );
      expect(asFragment()).toMatchSnapshot();
  });
});

describe("active item changes successfully", () => {
  xit('active index increases by 1 when "next" button is clicked', () => {
    const { asFragment } = render(
      <Provider>
        <Carousel gap={50}>
          <DemoComp index={1} />
          <DemoComp index={2} />
          <DemoComp index={3} />
          <DemoComp index={4} />
          <DemoComp index={5} />
          <DemoComp index={6} />
          <DemoComp index={7} />
          <DemoComp index={8} />
          <DemoComp index={9} />
          <DemoComp index={10} />
        </Carousel>
        <LeftButton />
        <RightButton />
      </Provider>
    )
    expect(asFragment()).toMatchSnapshot();
  });
});
