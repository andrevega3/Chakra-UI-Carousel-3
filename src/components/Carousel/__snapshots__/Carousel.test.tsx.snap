// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly Carousel renders correctly 1`] = `
<DocumentFragment>
  <div
    class="css-10uq0s4"
  >
    <div
      class="chakra-stack css-7qd5oz"
    >
      <div
        class="css-1kuahdq"
        draggable="false"
        style="transform: none; user-select: none;"
      >
        <div
          class="css-6vcq0v"
        >
          <div
            class="css-gi6qt2"
            id="carousel-item-1"
          >
            <h2
              class="chakra-heading css-4z9z2s"
            >
              Title 1
            </h2>
            <div
              class="css-gg4vpm"
            >
              <p
                class="css-bjcoli"
              >
                Body 1
              </p>
              <p
                class="css-a2wsy3"
              >
                user 1
              </p>
            </div>
          </div>
        </div>
        <div
          class="css-6vcq0v"
        >
          <div
            class="css-gi6qt2"
            id="carousel-item-2"
          >
            <h2
              class="chakra-heading css-4z9z2s"
            >
              Title 2
            </h2>
            <div
              class="css-gg4vpm"
            >
              <p
                class="css-bjcoli"
              >
                Body 2
              </p>
              <p
                class="css-a2wsy3"
              >
                user 2
              </p>
            </div>
          </div>
        </div>
        <div
          class="css-6vcq0v"
        >
          <div
            class="css-gi6qt2"
            id="carousel-item-3"
          >
            <h2
              class="chakra-heading css-4z9z2s"
            >
              Title 3
            </h2>
            <div
              class="css-gg4vpm"
            >
              <p
                class="css-bjcoli"
              >
                Body 3
              </p>
              <p
                class="css-a2wsy3"
              >
                user 3
              </p>
            </div>
          </div>
        </div>
        <div
          class="css-6vcq0v"
        >
          <div
            class="css-gi6qt2"
            id="carousel-item-4"
          >
            <h2
              class="chakra-heading css-4z9z2s"
            >
              Title 4
            </h2>
            <div
              class="css-gg4vpm"
            >
              <p
                class="css-bjcoli"
              >
                Body 4
              </p>
              <p
                class="css-a2wsy3"
              >
                user 4
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
